gc/base/standard/ConcurrentGCIncrementalUpdate.cpp

```diff
@@ -1,205 +1,1435 @@
+	{
+		_cardTable = MM_ConcurrentCardTable::newInstance(env, _extensions->getHeap(), _markingScheme, this);
+	}
+
+	if (NULL != _cardTable) {
+		result = true;
+		/* Set card table address in GC Extensions */
+		_extensions->cardTable = _cardTable;
+	}
+
+	return result;
+}
+
+#if defined(OMR_GC_MODRON_SCAVENGER)
+/**
+ * Process event from an external GC (Scavenger) when old-to-old reference is created
+ * from an existing old-to-new or new-to-new reference.
+ * We need to rescan the parent object, as slots which during original scan contained
+ * new area references could now contain references into old area.
+ * @param objectPtr  Parent old object that has a reference to a child old object
+ */
+void
+MM_ConcurrentGCIncrementalUpdate::oldToOldReferenceCreated(MM_EnvironmentBase *env, omrobjectptr_t objectPtr)
+{
+	/* todo: Consider doing mark and push-to-scan on child object.
+	 * Child object will be tenured only once during a Scavenge cycle,
+	 * so there are no risks of creating duplicates in scan queue. */
+	Assert_MM_true(CONCURRENT_OFF != _stats.getExecutionMode());
+	Assert_MM_true(_extensions->isOld(objectPtr));
+	if (_markingScheme->isMarkedOutline(objectPtr) ) {
+		_cardTable->dirtyCard(env,objectPtr);
+	}
+}
+#endif /* OMR_GC_MODRON_SCAVENGER */
+
+void
+MM_ConcurrentGCIncrementalUpdate::finalConcurrentPrecollect(MM_EnvironmentBase *env)
+{
+	OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
+
+	reportConcurrentFinalCardCleaningStart(env);
+	uint64_t startTime = omrtime_hires_clock();
+
+	bool overflow = false; /* assume the worst case*/
+	uintptr_t overflowCount;
+
+	do {
+		/* remember count when we start */
+		overflowCount = _stats.getConcurrentWorkStackOverflowCount();
+
+		/* Get assistance from all worker threads to do final card cleaning */
+		MM_ConcurrentFinalCleanCardsTask cleanCardsTask(env, _dispatcher, this, env->_cycleState);
+		((MM_ConcurrentCardTable *)_cardTable)->initializeFinalCardCleaning(env);
+
+		_dispatcher->run(env, &cleanCardsTask);
+
+		/* Have we had a work stack overflow whilst processing card table ? */
+		overflow = (overflowCount != _stats.getConcurrentWorkStackOverflowCount());
+	} while (overflow);
+
+	/* reset overflow flag */
+	_markingScheme->getWorkPackets()->clearOverflowFlag();
+
+	reportConcurrentFinalCardCleaningEnd(env, omrtime_hires_clock() - startTime);
+#if defined(DEBUG)
+	Assert_MM_true(_cardTable->isCardTableEmpty(env));
+#endif
+}
+
+/**
+ * Tune the concurrent adaptive parameters.
+ * Using historical data attempt to predict how much work (both tracing and
+ * card cleaning) will need to be performed during the next concurrent mark cycle.
+ */
+void
+MM_ConcurrentGCIncrementalUpdate::tuneToHeap(MM_EnvironmentBase *env)
+{
+	MM_Heap *heap = (MM_Heap *)_extensions->heap;
+	uintptr_t heapSize = heap->getActiveMemorySize(MEMORY_TYPE_OLD);
+	uintptr_t kickoffThreshold;
+	uintptr_t cardCleaningThreshold;
+	uintptr_t kickoffThresholdPlusBuffer;
+	uintptr_t totalBytesToTrace, totalBytesToClean;
```

Comment:
hanging variable declarations (uninitialized and/or declared too early) are pretty common in this code. do one pass and fix for all that you find.